src/target/riscv/riscv.c

```diff
@@ -2456,11 +2467,14 @@ static int riscv_read_memory(struct target *target, target_addr_t address,
 	}
 
 	target_addr_t physical_addr;
-	if (target->type->virt2phys(target, address, &physical_addr) == ERROR_OK)
-		address = physical_addr;
+	int result = target->type->virt2phys(target, address, &physical_addr);
+	if (result != ERROR_OK) {
+		LOG_TARGET_ERROR(target, "Address translation failed.");
+		return result;
+	}
 
 	RISCV_INFO(r);
-	return r->read_memory(target, address, size, count, buffer, size);
+	return r->read_memory(target, physical_addr, size, count, buffer, size);
```

Comment:
Wow. That's kind of a big bug. This code never worked at all?

Reply:
it's kinda worked. there was re-assignment of the address variable to a translated address.

Reply:
Yes, now we store real address in `physical_addr` and it filled by `riscv_virt2phys` even if MMU is disabled.